### PR TITLE
Final delivery

### PR DESCRIPTION
**Pull request by bornehill**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,4 +6,6 @@ require (
 	github.com/davecgh/go-spew v1.1.1
 	github.com/gorilla/mux v1.8.0
 	github.com/spf13/viper v1.7.1
+	github.com/stretchr/testify v1.3.0
+	github.com/vektra/mockery/v2 v2.4.0 // indirect
```
Comment (RolloCasanova) [inline go.mod:10]: It seems this is not used in the project, please remove it.

File: infrastructure/router/app_router.go [hunk 2]
```diff
@@ -10,10 +10,13 @@ type appRouter struct {
 	App controller.AppController
 }
 
+//AppRouter App router Interface
 type AppRouter interface {
 	GetCentres(w http.ResponseWriter, r *http.Request)
 }
 
+/*NewAppRouter Create new App controller Router
+<param name="ac">App Controller</param> */
```
Comment (RolloCasanova) [inline infrastructure/router/app_router.go:19]: I suppose this is used by some tool, may I ask what is this comment for?

File: infrastructure/router/app_router_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		},
+	}
+	for _, tt := range tests {
+		tt.ar.On("GetCentres", tt.args.w, tt.args.r).Return()
+
+		t.Run(tt.name, func(t *testing.T) {
+			tt.ar.GetCentres(tt.args.w, tt.args.r)
+		})
+	}
+}
```
Comment (RolloCasanova) [inline infrastructure/router/app_router_test.go:58]: Technically speaking, this is not testing anything.

You are mocking the `GetCentres` method in the router package, when you should be mocking the dependencies.

Also, setting w as nil (first test scenario), will result as panic, as it's a nil reference, and `router.GenCentres` method sets a header value.

File: infrastructure/router/app_router_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	tests := []struct {
+		name string
+		ar   *mocks.AppRouter
+		args args
+	}{
+		{
+			name: "Should call GetCentres",
+			ar:   &mocks.AppRouter{},
+			args: args{w: nil, r: &http.Request{}},
+		},
```
Comment (RolloCasanova) [inline infrastructure/router/app_router_test.go:49]: This scenario is not valid for this test. You should create scenarios for that specific`GetCentres` method call.

Such as:
- App.Center.GetCentres() success
- App.Center.GetCentres() error

And add validations in each one of these inside the test's loop; first, add an expected status response as part of the test, and compare (assert) that expected value is the one you are getting when calling your dependency (`ar.App.Centre.GenCentres()`.

EDIT.
Actually, nothing is being tested at all. Search for a code coverage tool to see what lines you are testing and you will see it with your own eyes.

We could discuss this in a meeting.

File: interface/presenter/centre_presenter_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("centrePresenter.JsonResponse() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func testData() (*[]*model.Centre, []byte) {
+	ret := make([]*model.Centre, 0, 0)
```
Comment (RolloCasanova) [inline interface/presenter/centre_presenter_test.go:79]: Better define a default size, when you set any value, and you have already on the maximum capacity, a reallocation happens and wasting resources (both time and space). 

Not a big deal as it's only two of them, but better know this beforehand.

**Review by RolloCasanova — COMMENTED** [review]
Let's keep working on the unit-testing; you are missing the key point of the unit-testing and mocking the dependencies. We can work on that if you want.

Let's also add some project information to the project README file.

Besides that, you improved from your first deliverable, and you have the Go basics already dominated.

Let's continue working on getting better at the language gotchas.

Welcome to the world of Go developers 👍.

File: infrastructure/router/app_router.go (hunk 2 above)
Comment (bornehill) [reply infrastructure/router/app_router.go:19]: It is a Lint warning for public methods

image
